python/ray/util/client/server/dataservicer.py

```diff
@@ -67,19 +84,25 @@ def Datapath(self, request_iterator, context):
         except grpc.RpcError as e:
             logger.debug(f"Closing data channel: {e}")
         finally:
-            logger.info(f"Lost data connection from client {client_id}")
+            logger.debug(f"Lost data connection from client {client_id}")
             self.basic_service.release_all(client_id)
 
-            with self._clients_lock:
-                self._num_clients -= 1
+            with self.clients_lock:
+                if accepted_connection:
+                    # Could fail before client accounting happens
+                    self.num_clients -= 1
+                    logger.debug(f"Removed clients. {self.num_clients}")
 
-            with disable_client_hook():
-                if self._num_clients == 0:
-                    ray.shutdown()
+                # It's important to keep the Ray shutdown
+                # within this locked context or else Ray could hang.
+                with disable_client_hook():
+                    if self.num_clients == 0:
+                        logger.debug("Shutting down ray.")
+                        ray.shutdown()
```

Comment:
Can you add a unit test for the max client limit (see test_client_Init.py)?

Reply:
done